server/form-pages/apply/about-the-person/address-history/previousAddress.test.ts

```diff
@@ -0,0 +1,137 @@
+      describe('when there are no errors', () => {
+        it('returns empty object', () => {
+          const page = new PreviousAddress(previousAddress, application)
+          expect(page.errors()).toEqual({})
+        })
+      })
+      const requiredFields = [
+        ['previousAddressLine1', 'Enter the first line of the address'],
+        ['previousTownOrCity', 'Enter town or city'],
+        ['previousPostcode', 'Enter a postcode'],
+      ]
+
+      it.each(requiredFields)('it includes a validation error for %s', (field, message) => {
+        const page = new PreviousAddress({ hasPreviousAddress: 'yes' }, application)
+        const errors = page.errors()
+
+        expect(errors[field]).toEqual(message)
+      })
+    })
+
+    describe('when there is not a previous address', () => {
+      describe('when there are no errors', () => {
+        it('returns empty object', () => {
+          const page = new PreviousAddress(lastKnownAddress, application)
+          expect(page.errors()).toEqual({})
+        })
+      })
+      const requiredFields = [
+        ['lastKnownAddressLine1', 'Enter the first line of the address'],
+        ['lastKnownTownOrCity', 'Enter town or city'],
+        ['lastKnownPostcode', 'Enter a postcode'],
+      ]
+
+      it.each(requiredFields)('it includes a validation error for %s', (field, message) => {
+        const page = new PreviousAddress({ hasPreviousAddress: 'no' }, application)
+        const errors = page.errors()
+
+        expect(errors[field]).toEqual(message)
+      })
+    })
+  })
+
+  describe('response', () => {
+    it('returns empty object', () => {
+      const applicationWithNoData = { ...application, data: undefined } as Cas2Application
+      const page = new PreviousAddress({}, applicationWithNoData)
+      expect(page.response()).toEqual({})
+    })
+
+    describe('when there is a previous address', () => {
+      const applicationWithData = {
+        ...application,
+        data: { 'address-history': { 'previous-address': { ...previousAddress } } },
+      }
+      const page = new PreviousAddress({}, applicationWithData)
+      expect(page.response()).toEqual({
+        'Did Roger Smith have an address before entering custody?': 'Yes',
+        'What was the address?': `1 Example Road\r\nPretend Close\r\nAberdeen\r\nGloucestershire\r\nAB1 2CD\r\n`,
+      })
+    })
+
+    describe('when there no previous address', () => {
+      const applicationWithData = {
+        ...application,
+        data: { 'address-history': { 'previous-address': { ...lastKnownAddress } } },
+      }
+      const page = new PreviousAddress({}, applicationWithData)
+      expect(page.response()).toEqual({
+        'Did Roger Smith have an address before entering custody?': 'No fixed address',
+        'How long did the person have no fixed address for?': '6 months',
+        'What was the address?': `2 Example Road\r\n2 Pretend Close\r\nBristol\r\nShropshire\r\nEF1 GHD\r\n`,
```

Comment:
In this case, surely we want the question to be 'What was Roger Smith's last known permanent address?'

Reply:
hmm, I'm not sure, because this is also our audit trail for exactly what was asked and what was answered. I'm worried about making choices like that the hide the exact questions...

Reply:
ohh I totally misunderstood your question! Very good spot, Chanel picked it up as well and I'm updating